fdbserver/worker.actor.cpp

```diff
@@ -748,17 +749,61 @@ ACTOR Future<Void> workerSnapCreate(WorkerSnapRequest snapReq, StringRef snapFol
 	return Void();
 }
 
-ACTOR Future<Void> monitorServerDBInfo( Reference<AsyncVar<Optional<ClusterControllerFullInterface>>> ccInterface, Reference<ClusterConnectionFile> connFile, LocalityData locality, Reference<AsyncVar<ServerDBInfo>> dbInfo ) {
+ACTOR Future<Void> monitorTraceLogIssues(Optional<Reference<AsyncVar<std::set<std::string>>>> issues) {
+	state bool pingTimeout = false;
+	loop {
+		wait(delay(SERVER_KNOBS->TRACE_LOG_FLUSH_FAILURE_CHECK_INTERVAL_SECONDS));
+		ThreadFuture<Void> f(new ThreadSingleAssignmentVar<Void>);
+		Reference<CompletionCallback<Void>> callback =
+		    Reference<CompletionCallback<Void>>(new CompletionCallback<Void>(f));
+		callback->self = callback;
+		f.callOrSetAsCallback(callback.getPtr(), callback->userParam, 0);
+		pingTraceLogWriterThread(f);
+		try {
+			wait(timeoutError(callback->promise.getFuture(), SERVER_KNOBS->TRACE_LOG_PING_TIMEOUT_SECONDS));
+		} catch (Error& e) {
+			if (e.code() == error_code_timed_out) {
+				pingTimeout = true;
+			} else {
+				throw;
+			}
+		}
+		if (issues.present()) {
+			std::set<std::string> _issues = getTraceLogIssues();
+			if (pingTimeout) {
+				// Ping trace log writer thread timeout.
+				_issues.insert("trace_log_writer_thread_unresponsive");
+				pingTimeout = false;
+			}
+			if (_issues.size() > 0) {
+				issues.get()->set(_issues);
+			}
+		}
+	}
+}
+
+ACTOR Future<Void> monitorServerDBInfo(Reference<AsyncVar<Optional<ClusterControllerFullInterface>>> ccInterface,
+                                       Reference<ClusterConnectionFile> connFile, LocalityData locality,
+                                       Reference<AsyncVar<ServerDBInfo>> dbInfo,
+                                       Optional<Reference<AsyncVar<std::set<std::string>>>> issues) {
 	// Initially most of the serverDBInfo is not known, but we know our locality right away
 	ServerDBInfo localInfo;
 	localInfo.myLocality = locality;
 	dbInfo->set(localInfo);
 
 	state Optional<double> incorrectTime;
+	state bool checkIssues = false;
 	loop {
 		GetServerDBInfoRequest req;
 		req.knownServerInfoID = dbInfo->get().id;
 
+		if (issues.present() && checkIssues) {
+			for (auto const& i : issues.get()->get()) {
+				req.issues.push_back_deep(req.issues.arena(), i);
+			}
+			checkIssues = false;
```

Comment:
Will this result in issues only being pushed once, even if we do multiple of these requests? I think we would need the issue set to be sent each time, but my understanding may be incorrect.

Reply:
Can you elaborate the case where there will be multiple these requests?

Reply:
Oh I think I got your point. The issues will be, wrongly, reported to be cleared on subsequent requests after the one triggered by `issues.get()->onChange()`. 

One solution is that I can remove the `if (_issues.size() > 0)` condition above in the period issues checking actor, which I think will cause the loop here inside `monitorServerDBInfo` to run every `TRACE_LOG_FLUSH_FAILURE_CHECK_INTERVAL_SECONDS` seconds. Then this will require the issues to be explicitly cleared, which is basically what your other comment suggests.

Reply:
Thanks pointing this out!

Reply:
Fixed.